pocs/utils/database.py

```diff
@@ -111,295 +229,278 @@ def __init__(self,
+                # contain any characters that json would need to escape: first
+                # check if the line contains the obj_id; if not skip. Else, parse
+                # as json, and then check for the _id match.
+                obj = json_util.loads(line)
+                if obj['_id'] == obj_id:
+                    return obj
+        return None
 
     def clear_current(self, type):
-        """Clear the current record of a certain type
-
-        Args:
-            type (str): The type of entry in the current collection that
-                should be cleared.
-        """
         current_f = os.path.join(self._storage_dir, 'current_{}.json'.format(type))
         try:
             os.remove(current_f)
         except FileNotFoundError as e:
             pass
 
+    def _get_file(self, collection, permanent=True):
+        if permanent:
+            name = '{}.json'.format(collection)
+        else:
+            name = 'current_{}.json'.format(collection)
+        return os.path.join(self._storage_dir, name)
+
     def _make_id(self):
         return str(uuid4())
 
-    def _warn(self, *args, **kwargs):
-        if hasattr(self, 'logger'):
-            self.logger.warning(*args, **kwargs)
-        else:
-            warn(*args)
+    @classmethod
+    def reset_for_test(cls, db_name):
+        # Clear out any .json files.
+        storage_dir = os.path.join(os.environ['PANDIR'], 'json_store', db_name)
+        for f in glob(os.path.join(storage_dir, '*.json')):
+            os.remove(f)
+
+
+class PanMemoryDB(AbstractPanDB):
+    """In-memory store of serialized objects.
+
+    We serialize the objects in order to test the same code path used
+    when storing in an external database.
+    """
+
+    active_dbs = weakref.WeakValueDictionary()
+
+    @classmethod
+    def get_or_create(cls, db_name=None, **kwargs):
+        """Returns the named db, creating if needed.the
+
+        This method exists because PanDB gets called multiple times for
+        the same database name, With mongo or a file store where the storage
+        is external from the instance, that is not a problem, but with
+        PanMemoryDB the instance is the store, so the instance must be
+        shared."""
+        db = PanMemoryDB.active_dbs.get(db_name)
+        if not db:
+            db = PanMemoryDB(db_name=db_name, **kwargs)
+            PanMemoryDB.active_dbs[db_name] = db
+        return db
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.current = {}
+        self.collections = {}
+        self.lock = threading.Lock()
+
+        logger = self.logger or get_root_logger()
+        logger.info('Create new PanMemoryDB for {}', self.db_name)
+
+    def _make_id(self):
+        return str(uuid4())
+
+    def insert_current(self, collection, obj, store_permanently=True):
+        self._warn('insert_current {!r} {!r}'.format(collection, obj))
```

Comment:
Do you want to keep this in here in the long run or was this mostly for testing?